### ISSUE TITLE
Review + Expand Instructions

### ISSUE THREAD

**Issue by nelsonic** [dwyl/phoenix-liveview-chat-example#42]
## Context

We are using `LiveView` for a Client project. 
The project already has some basic `LiveView` UI but has barely scratched the surface of what's possible.
Specifically I want to add "who is logged in" (presence) to our `gitea-demo` App's "advanced" example: https://github.com/dwyl/gitea-demo/issues/12

This `LiveView` _Chat_ Example App is already an _excellent_ starting point for understanding how to do this.
But I feel that as a complete beginner a couple of steps can be expanded. 💭 

## Todo

+ [x] Go through all the steps on `localhost`
+ [x] Update/fill-in any gaps noted.
+ [x] _Understand_ why builds are failing for PRs: https://github.com/dwyl/phoenix-liveview-chat-example/pulls
![image](https://user-images.githubusercontent.com/194400/171508611-1fafafda-a789-436b-a73c-605e84ad07d1.png)
+ [ ] Create new Fly.io org to deploy the app so we have an instant-loading demo
  + [ ] Invite @SimonLab (original author) to be `admin` of the new org. 
  > having a new org means the app runs in "free tier" ...)
+ [x] Use https://auth.dwyl.com (Fly.io) as the auth provider as it's _much_ faster than https://dwylauth.herokuapp.com
+ [x] Bonus: add style #4 

Going to spend **`T4h`** on this tomorrow morning as I want to reference it in the work I'm doing in https://github.com/dwyl/gitea-demo/issues/12

Will assign PR once ready. 👍

**Comment by nelsonic** [comment]
Took a minor detour to investigate using Tailwind ... https://fly.io/phoenix-files/tailwind-standalone/ 💭

**Comment by nelsonic** [comment]
Technology stacks are like Automobile Brands. There are _dozens_ of them. 
None are "right" or "wrong". They all have pros and cons. Costs and benefits. 

![image](https://user-images.githubusercontent.com/194400/156757708-082f909a-9f38-498e-bd9d-2b07e6476f09.png)

> **Note**: This graphic illustrates the "big" manufacturers/brands 
and doesn't include brands you have definitely heard of such as Mazda, Subaru or Susuki 
or smaller/niche brands like Ferrari, McLaren, Morgan and Koenigsegg
or newer brands like BYD, Fisker, Lucid, NIO, XPeng, Rivian & Tesla ... 
Bottom line: there are _many_ ways of building vehicles. 
Equally there are _many_ ways to build software and especially web-based software.

**Comment by SimonLab** [comment]
Looking at the failing tests, the error is linked to `auth_plug` where the `api_key` environment variable is not defined

**Comment by nelsonic** [comment]
The build passed on Simon's PR #45 
But fails on `main`:
https://github.com/dwyl/phoenix-liveview-chat-example/runs/6757621874?check_suite_focus=true#step:7:140
![image](https://user-images.githubusercontent.com/194400/172361859-b1f09cb8-af3e-42b5-8251-ce37707a94ee.png)

going to try and re-run it just to see if it's an _intermittent_ failure or perm one. ♻️

**Comment by nelsonic** [comment]
Still fails: 
![image](https://user-images.githubusercontent.com/194400/172362390-b9187d82-635e-432b-a65d-832e6b5581e1.png)

https://github.com/dwyl/phoenix-liveview-chat-example/runs/6772512125?check_suite_focus=true#step:7:137

![image](https://user-images.githubusercontent.com/194400/172362484-8beeb245-e696-45d2-9cc9-b83b80800e54.png)

Lame. Looks like an issue in `AuthPlug` we can investigate this _later_. 🤞

**Comment by SimonLab** [comment]
The timeout issue is due to the auth Heroku app not being awake for the login test I think. We could mock this call to Auth maybe

**Comment by nelsonic** [comment]
@SimonLab yeah, I think our `Auth` an `AuthPlug` code needs an overhaul. 💭
In the meantime do you think we could use https://auth.dwyl.com/ (Fly.io) instead of Heroku for the auth in _this_ App so that there's lower chance of timeout? ⌛ 

> I'm thinking we should build a **`v2.0`** of **`Auth`** using [**`mix phx.gen.auth`**](https://github.com/dwyl/auth/issues/133#issuecomment-1076169035) as the starting point
but with **`Tailwind`** for the Style/UI etc. ✨ 
See: https://github.com/dwyl/auth/issues/207 💡 
How busy are you next week? 😜 
All joking asside. We should catch-up when you're free! 🙏

**Comment by nelsonic** [comment]
```sh
Could not render "messages.html" for LiveviewChatWeb.MessageView, 
please define a matching clause for render/2 
or define a template at "lib/liveview_chat_web/templates/message/*". 

The following templates were compiled:

* message.html
```

![image](https://user-images.githubusercontent.com/194400/172559435-9888f370-8e93-4fb9-8007-922e6d86751e.png)

subtle ...

_Current_ instructions: 
https://github.com/dwyl/phoenix-liveview-chat-example/tree/4857be73ed4d7e27426357d18621afad4aa1f197#liveview-route-controller-and-template

had:
```elixir
LiveviewChatWeb.MessageView.render("messages.html", assigns)
```

Should be:
```elixir
LiveviewChatWeb.MessageView.render("message.html", assigns)
```

i.e. the template file is `message.html.heex` but the `render/2` was trying to render `messages...` (plural!)
This is could either be a _useful_ debugging exercise for the person following along, or a frustrating roadblock.

Anyway, updated it. Working. ✅ 

![image](https://user-images.githubusercontent.com/194400/172560880-86e92751-2c00-4daf-9e6a-b428dec344ea.png)

**Comment by nelsonic** [comment]
```sh
Resolving Hex dependencies...
The dependency resolver is taking more than 30 seconds. This typically happens when Hex cannot find a suitable set of dependencies that match your requirements. Here are some suggestions:

  1. Do not delete mix.lock. If you want to update some dependencies, do mix deps.update dep1 dep2 dep3

  2. Tighten up your dependency requirements to the latest version. Instead of {:my_dep, ">= 1.0.0"}, try {:my_dep, "~> 3.6"}
```
# ⏳

**Comment by nelsonic** [comment]
It's not my internet speed ... 🤷‍♂️ 

![image](https://user-images.githubusercontent.com/194400/172576854-d01e140f-c7d0-4d2c-861d-330ceba6fc87.png)

**Comment by nelsonic** [comment]
Took a minor detour to learn **`Alpine.js`**: https://github.com/dwyl/learn-alpine.js 🚀 
Going to use it and **`Tailwind CSS`** for the UI enhancements #4 😉

**Comment by nelsonic** [comment]
This is not a particularly beginner-friendly error message/page: 🙄 

![image](https://user-images.githubusercontent.com/194400/174010187-f36835e4-409d-4ee4-a4d6-9ec8d774e91f.png)

Where does someone new to `Phoenix` / `LiveView` even _begin_ to debug this? 🤷‍♂️ 
It looks like an _internal_ error, not being allowed to set the `id` of a DOM attribute to `1` ... 
This is because we are using the `message.id` as the `id` in the UI for uniqueness.

Luckily it's on `localhost` and I just nuked the DB `mix ecto.reset` and it worked!

![image](https://user-images.githubusercontent.com/194400/174010587-b0641eea-c9d2-41eb-8261-c7fa12f34912.png)

But if I were trying to follow along with the tutorial and saw this I would be pretty discouraged!

**Comment by nelsonic** [comment]
The offending line of code is:
```elixir
<li id={message.id}>
```
in: https://github.com/dwyl/phoenix-liveview-chat-example/blame/4857be73ed4d7e27426357d18621afad4aa1f197/README.md#L233

It's correct in the file:

https://github.com/dwyl/phoenix-liveview-chat-example/blob/4857be73ed4d7e27426357d18621afad4aa1f197/lib/liveview_chat_web/templates/message/message.html.heex#L3

But not in the `README.md` instructions.

**Comment by nelsonic** [comment]
Gif at end of section 9:
![liveview-chat-demo](https://user-images.githubusercontent.com/194400/174016930-52b73247-eb7e-4c3e-8a4d-db0929aacc39.gif)

**Comment by nelsonic** [comment]
Going to leave this on Heroku for now. We can deploy to Fly.io in a follow-up PR. 💭

**Comment by nelsonic** [comment]
4 actions:
<img width="692" alt="image" src="https://user-images.githubusercontent.com/194400/174164940-9d27689e-022d-40c1-8be0-60e915256b19.png">

Just to request review of a task.
We can do _much_ better!

**Comment by nelsonic** [comment]
PR: https://github.com/dwyl/phoenix-liveview-chat-example/pull/46

**Comment by nelsonic** [comment]
Heroku App not auto-deploying: #47 🙃
